src/value.rs

```diff
@@ -156,26 +163,36 @@ impl Value {
         }
     }
 
-    /// Tries to get a value of type `T`.
-    ///
-    /// Returns `Some` if the type is correct and the value is not `None`.
+    /// Tries to get a possibly optional value of type `T`.
     ///
-    /// This function doesn't distinguish between type mismatches and correctly
-    /// typed `None` values. Use `downcast` or `is` for that.
-    pub fn get<'a, T: FromValueOptional<'a>>(&'a self) -> Option<T> {
+    /// Returns `Ok` if the type is correct.
+    pub fn get<'a, T: FromValueOptional<'a>>(&'a self) -> Result<Option<T>, BoolError> {
         unsafe {
             let ok = from_glib(gobject_sys::g_type_check_value_holds(
                 mut_override(self.to_glib_none().0),
                 T::static_type().to_glib(),
             ));
             if ok {
-                T::from_value_optional(self)
+                Ok(T::from_value_optional(self))
             } else {
-                None
+                Err(glib_bool_error!(format!(
+                    "Value type mismatch (actual: {:?}, requested: {:?})",
+                    self.type_(),
+                    T::static_type(),
+                )))
             }
         }
     }
 
+    /// Tries to get a value of type `T`.
+    ///
+    /// Returns `Ok` if the type is correct and `Some` value is available.
+    /// (check the `BoolError` description for details in case of an error).
+    pub fn get_some<'a, T: FromValueOptional<'a>>(&'a self) -> Result<T, BoolError> {
```

Comment:
This is a weird one :) I think you want the trait bound to be simply `FromValue`, just like `TypeValue::get_some()` works. For types implementing that trait there can't be a `None`

Reply:
I thought it would also be useful when you expect a value to be defined, even though the type is "option-able".

Reply:
That seems wrong. In that case you'd explicitly unwrap/expect because *you* know that the value exists :) Deferring that to the function here hides this assumption in the code

Reply:
It doesn't seem hidden to me since it is explicit in the function name :)

If we were expected to use it solely for `FromValue` types, we should name it something like `get_non_optional` (I agree this is ugly) :)

Reply:
I'd keep `get_some()` consistent with what we have right now in `TypedValue`. So either keep the name and change this to `FromValue` here, or rename both and have them both work with `FromValue` :)

Reply:
I can do that!

@GuillaumeGomez & @EPashkin, what do you think?

Reply:
I'm not sure we want to have something "option-able" in such a generic context. `FromValue` seems good enough for me.

Reply:
But keep it called `get_some()` (which is weird), or name it differently (how?)?

Reply:
I don't have opinion on this point. Good luck! :)

Reply:
`get_non_optional()` seems most descriptive to me, just a bit long to type. I don't have any better suggestions :)

Reply:
`get_mandatory`? But that not accurate IMO as the types are not optional anyway.

Another option would be to rename current `get` as `get_optional` and to implement `FromValueOptional` only on types that can host `None`. The trait would be renamed as `FromOptionalValue` and we could rename current `get_some` methods as `get`.

Reply:
> Another option would be to rename current `get` as `get_optional`

That seems fine

> The trait would be renamed as `FromOptionalValue` and we could rename current `get_some` methods as `get`.

That too but why rename?

> Implement `FromValueOptional` only on types that can host `None`

And this would break things :) The optional types are a full superset of the non-optional ones. You can handle each non-optional type (e.g. the integers) as a optional one, you'd just get a `Some` around it, always.

Why would you change this?

----

My main worry with changing `get()` to only work on non-optional types is all the user complaints. Most of the time you work with optional types and `get()` is much more visible than some other `get_XXX()`, and people will probably get confused. And they'll need to update more of their code than otherwise.

Reply:
I was proposing to rename `FromValueOptional` to reflect that it would only apply to optional values (as opposed to yielding an `Option` from a `Value`). I noticed this trait was a superset of `FromValue` type, but it seemed confusing IMHO. Why would we want to get an `Option<T>` when we explicitly indicate a type that is not an option-able? Then I though that `FromValueOptional` was implemented for these types for convenience when used in `get` methods. Using `get` / `get_optional` would make it no longer necessary to be a superset. But I must admit I didn't engage in a large scale survey, so I might be missing other use cases.

I understand your concerns about users being bothered by these changes. If we decide to proceed with these changes, I could add a comment in the documentation of the new `get` methods and link to the `get_optional` versions to clarify. I think it is convenient to being able to `get` embedded numerics, booleans and types without the extra `unwrap()`.

Reply:
Having the `get()` work for all types allows you to write generic code over all types, instead of having to special case both traits everywhere.

Reply:
Ok, so I'll keep the `get` methods as-is and will try to figure out a better name for `get_some`, possibly `get_non_optional`.

Reply:
Ok! Not great either, I agree. But all alternatives I can see currently seem worse :)

Reply:
Looks good to me now otherwise, only needs renaming I guess